src/predicates.ts

```diff
@@ -0,0 +1,596 @@
+  }
+
+  public test(configOption: ConfigOption): RuleMatchedAt {
+
+    const patternOption: string = `ConfigOption "${configOption.name}"`;
+    const patternOptionEquals: string = `ConfigOptionEquals "${configOption.name}"`;
+
+    const configOptionUsages: FoundValue[] = _.concat(
+        this.valueSearcher.searchContains(configOption, patternOption),
+        this.valueSearcher.searchContains(configOption, patternOptionEquals),
+    );
+
+    return {
+      matched: !!configOptionUsages.length,
+      paths: _.map(configOptionUsages, f => f.path),
+    };
+  }
+}
+
+interface InvalidMontior {
+  index: number;
+  path: string;
+}
+
+export class MonitorContainerMissing implements Predicate<any> {
+  public static fromJson(self: any): MonitorContainerMissing {
+    return new MonitorContainerMissing(self.monitorPath);
+  }
+
+  constructor(private readonly monitorPath: string) {
+  }
+
+  public test(root: any): RuleMatchedAt {
+    if (!_.has(root, this.monitorPath)) {
+      return { matched: false };
+    }
+
+    const monitors: string[] = _.get(root, this.monitorPath) as string[];
+
+    if (_.isEmpty(monitors)) {
+      return { matched: false };
+    }
+
+    if (_.isEmpty(_.get(root, "components"))) {
+      return { matched: true, paths: [this.monitorPath] };
+    }
+
+    const violations: InvalidMontior[] = _.filter(_.map(monitors, (monitor, index) => {
+      return this.checkMonitor(root.components, monitor, index);
+    })) as any;
+
+    if (_.isEmpty(violations)) {
+      return { matched: false };
+    }
+
+    return {
+      matched: true,
+      paths: [
+        ..._.map(violations, v => `${this.monitorPath}.${v.index}`),
+        ..._.map(violations, v => v.path),
+      ],
+    };
+  }
+
+  private checkMonitor(components: any[], monitor: string, index: number): InvalidMontior | undefined {
+    const [name, image] = monitor.split(",");
+
+    // fail if component missing
+    const componentIndex: any = _.findIndex(components, { name });
+    if (componentIndex === -1) {
+      return { index, path: "components" };
+    }
+
+    // fail if container missing
+    const container = _.find(components[componentIndex].containers, { image_name: image });
+    if (!container) {
+      return { index, path: `components.${componentIndex}` };
+    }
+  }
+}
+
+export class SemverRange implements Predicate<any> {
+  public static fromJson(obj: any): SemverRange {
+    return new SemverRange(obj.path, obj.required);
+  }
+
+  constructor(
+      private readonly path: string,
+      private readonly required: boolean,
+  ) {
+  }
+
+  public test(root: any): RuleMatchedAt {
+    const value = _.get(root, this.path);
+
+    if (!value && this.required) {
+      return { matched: true, paths: [] };
+    }
+
+    if (value && !semver.validRange(value)) {
+      return { matched: true, paths: [this.path] };
+    }
+
+    return { matched: false };
+  }
+}
+
+export class AdminCommandContainerMissing implements Predicate<any> {
```

Comment:
this is the only new one, everything else in this file was just moved over from `engine.ts`